Storage/Logins.swift

```diff
@@ -0,0 +1,239 @@
+    }
+
+    init(credential: NSURLCredential, protectionSpace: NSURLProtectionSpace) {
+        self.credentials = credential
+        self.protectionSpace = protectionSpace
+    }
+
+    public func toDict() -> [String: String] {
+        return [
+            "hostname": hostname,
+            "formSubmitURL": formSubmitUrl ?? "",
+            "httpRealm": httpRealm ?? "",
+            "username": username ?? "",
+            "password": password,
+            "usernameField": usernameField ?? "",
+            "passwordField": passwordField ?? ""
+        ]
+    }
+
+    public class func fromScript(url: NSURL, script: [String: String]) -> LoginData {
+        let login = Login(hostname: getPasswordOrigin(url.absoluteString!)!, username: script["username"]!, password: script["password"]!)
+
+        if let formSubmit = script["formSubmitUrl"] {
+            login.formSubmitUrl = formSubmit
+        }
+
+        if let passwordField = script["passwordField"] {
+            login.passwordField = passwordField
+        }
+
+        if let userField = script["usernameField"] {
+            login.usernameField = userField
+        }
+
+        return login as LoginData
+    }
+
+    private class func getPasswordOrigin(uriString: String, allowJS: Bool = false) -> String? {
+        var realm: String? = nil
+        if let uri = NSURL(string: uriString) {
+            if allowJS && uri.scheme == "javascript" {
+                return "javascript:"
+            }
+
+            realm = "\(uri.scheme!)://\(uri.host!)"
+
+            // If the URI explicitly specified a port, only include it when
+            // it's not the default. (We never want "http://foo.com:80")
+            if var port = uri.port {
+                realm? += ":\(port)"
+            }
+        } else {
+            // bug 159484 - disallow url types that don't support a hostPort.
+            // (although we handle "javascript:..." as a special case above.)
+            println("Couldn't parse origin for \(uriString)")
+            realm = nil
+        }
+        return realm
+    }
+}
+
+public func ==(lhs: Login, rhs: Login) -> Bool {
+    return lhs.credentials == rhs.credentials && lhs.protectionSpace == rhs.protectionSpace
+}
+
+public protocol Logins {
+    func getUsageDataForLogin(login: LoginData) -> Deferred<Result<LoginUsageData>>
+    func getLoginsForProtectionSpace(protectionSpace: NSURLProtectionSpace) -> Deferred<Result<Cursor<LoginData>>>
+    func addLogin(login: LoginData) -> Success
+    func updateLogin(login: LoginData) -> Success
+    func addVisitFor(login: LoginData) -> Success
+    func removeLogin(login: LoginData) -> Success
+    func removeAll() -> Success
+}
+
+public class LoginDataError: ErrorType {
+    public let description: String
+    public init(description: String) {
+        self.description = description
+    }
+}
+
+public class MockLogins : Logins {
```

Comment:
We should probably shift this into a separate source file that's a member of the test targets.